HDFBatchEventsOutputer.cc

```diff
@@ -0,0 +1,356 @@
+  std::get<0>((*batchContainer)[indexInBatch]) = iEventID;
+  std::get<1>((*batchContainer)[indexInBatch]) = std::move(offsets);
+  std::get<2>((*batchContainer)[indexInBatch]) = std::move(buffer);
+
+  assert(waitingEventsInBatch_[batchIndex].load() < batchSize_);
+
+  auto waitingEvents = ++waitingEventsInBatch_[batchIndex];
+
+  if(waitingEvents == batchSize_ ) {
+    const_cast<HDFBatchEventsOutputer*>(this)->finishBatchAsync(batchIndex, std::move(iCallback));
+  }
+  auto time = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now() - start);
+  parallelTime_ += time.count();
+}
+
+void HDFBatchEventsOutputer::printSummary() const  {
+  //make sure last batches are out
+
+  auto start = std::chrono::high_resolution_clock::now();
+  {
+    tbb::task_group group;
+    
+    {
+      TaskHolder th(group, make_functor_task([](){}));
+      for( int index=0; index < waitingEventsInBatch_.size();++index) {
+        if(0 != waitingEventsInBatch_[index].load()) {
+          const_cast<HDFBatchEventsOutputer*>(this)->finishBatchAsync(index, th);
+        }
+      }
+    }
+    
+    group.wait();
+  }
+  auto writeTime = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now() - start);
+
+  std::cout <<"HDFBatchEventsOutputer\n  total serial time at end event: "<<serialTime_.count()<<"us\n"
+    "  total parallel time at end event: "<<parallelTime_.load()<<"us\n";
+  std::cout << "  end of job file write time: "<<writeTime.count()<<"us\n";
+
+  summarize_serializers(serializers_);
+}
+
+void HDFBatchEventsOutputer::finishBatchAsync(unsigned int iBatchIndex, TaskHolder iCallback) {
+
+  std::unique_ptr<std::vector<EventInfo>> batch(eventBatches_[iBatchIndex].exchange(nullptr));
+  auto eventsInBatch = waitingEventsInBatch_[iBatchIndex].load();
+  waitingEventsInBatch_[iBatchIndex] = 0;
+
+  std::vector<EventIdentifier> batchEventIDs;
+  batchEventIDs.reserve(batch->size());
+
+  std::vector<uint32_t> batchOffsets;
+  batchOffsets.reserve(batch->size() * (serializers_.size()+2));
+  //one extra size to but the final blob size. This is either the
+  // compressed size or the uncompressed size depending on the compression choice
+
+  std::vector<char> batchBlob;
+
+  int index = 0;
+  for(auto& event: *batch) {
+    if(index++ == eventsInBatch) {
+      //batch was smaller than usual. Can happen at end of job
+      break;
+    }
+    batchEventIDs.push_back(std::get<0>(event));
+
+    auto& offsets = std::get<1>(event);
+    std::copy(offsets.begin(), offsets.end(), std::back_inserter(batchOffsets));
+
+    auto& blob = std::get<2>(event);
+    //record the size of the blob as the final offset. Needed to decompress
+    // the event during reading
+    batchOffsets.push_back(blob.size());
+    std::copy(blob.begin(), blob.end(), std::back_inserter(batchBlob));
+
+    //release memory
+    blob = std::vector<char>();
+  }
```

Comment:
Some minor cleanups possible here.

```suggestion
  for(auto& [id, offsets, blob]: *batch) {
    if(index++ == eventsInBatch) {
      //batch was smaller than usual. Can happen at end of job
      break;
    }
    batchEventIDs.push_back(id);
    std::copy(offsets.begin(), offsets.end(), std::back_inserter(batchOffsets));
    //record the size of the blob as the final offset. Needed to decompress
    // the event during reading
    batchOffsets.push_back(blob.size());
    std::copy(blob.begin(), blob.end(), std::back_inserter(batchBlob));
 
    //release memory
    blob = {};
  }
```

Reply:
done